glyphs-reader/src/font.rs

```diff
@@ -768,26 +768,85 @@ fn default_master_idx(raw_font: &RawFont) -> usize {
         .unwrap_or(0)
 }
 
+fn axis_index(from: &RawFont, pred: impl Fn(&Axis) -> bool) -> Option<usize> {
+    from.axes
+        .as_ref()
+        .map(|axes| {
+            axes.iter()
+                .enumerate()
+                .find_map(|(i, a)| if pred(a) { Some(i) } else { None })
+        })
+        .unwrap_or_default()
+}
+
 fn extract_axis_mappings(from: &RawFont) -> BTreeMap<String, RawUserToDesignMapping> {
     let mut axis_mappings: BTreeMap<String, RawUserToDesignMapping> = BTreeMap::new();
+
+    // Fetch mapping from Axis Mappings, if any
     if let Some((_, axis_map)) = custom_param(&from.other_stuff, "Axis Mappings") {
         let Plist::Dictionary(axis_map) = axis_map.get("value").unwrap() else {
-            panic!("Incomprehensible axis map");
+            panic!("Incomprehensible axis map {:?}", axis_map);
         };
         for (axis_tag, mappings) in axis_map.iter() {
             let Plist::Dictionary(mappings) = mappings else {
-                panic!("Incomprehensible mappings");
+                panic!("Incomprehensible mappings {:?}", mappings);
+            };
+            let Some(axis_index) = axis_index(from, |a| &a.tag == axis_tag) else {
+                panic!("No such axes: {:?}", axis_tag);
             };
+            // We could not have found an axis index if there are no axes
+            let axis_name = &from.axes.as_ref().unwrap().get(axis_index).unwrap().name;
             for (user, design) in mappings.iter() {
                 let user: f32 = user.parse().unwrap();
                 let design = design.as_f64().unwrap() as f32;
                 axis_mappings
-                    .entry(axis_tag.clone())
+                    .entry(axis_name.clone())
                     .or_default()
                     .push((user.into(), design.into()));
             }
         }
     }
+
+    // Prefer to use Axis Location values if present
+    // https://github.com/googlefonts/fontmake-rs/issues/22 axis location gives user coord,
+    // the axis value gives the design coord.
+    for master in &from.font_master {
```

Comment:
you are only iterating over the font masters, but the Axis Location custom parameter can also appear on instances, and these also contribute to defining the axis mappings.

Reply:
And if an instance has the same internal design coordinate as a master, it also must share the same external/user-space axis location coordinate as that master, otherwise the two would be inconsistent.
And of course, the instances coords must be within the range of the masters' as the former are interpolated from the latter.

Reply:
I think I'd prefer to leave that for it's own PR and land this with just master support. Added a note to #22, unmarked this issue as fixing that one.